HAL.md

```diff
@@ -0,0 +1,100 @@
+# Hardware abstract layer(HAL) overview
+
+## Abstract
+Hardware abstract layer(HAL) defines a standard interface for hardware vendors to implement, which enables Tangle-Accelerator to be aware of lower-level driver implementations. HAL layer supports such as uart, GPIO, secure storage and other low-level operators inside endpoint. The endpoint is a daemon, offering monitor for the device and being able to send the message to tangle-accelerator. HAL layer lets the user only need to implement their device operators from abstract interface which is defined inside `device.h` but not need to know how endpoint works.
+
+## How to implement new device
+
+Create directory for the new device under `devices`.
+```
+mkdir devices/mydevice
+```
+* Create `impl.c` and `Makefile` under new device directory.
+* Include `device.h` headers into new device header or new device source.
+
+Here are some operations needed to implement inside new device:
+* device_operations
+    * init : initialize device
+    * fini : finalize device
+    * get_key : get device key
+    * get_device_id : get device id(IMEI or other identifier)
+* uart_operations
+    * init : initialize uart
+    * write : write command to uart device
+    * read : read from uart device
+    * clean : flush buffer
+* secure_store_operations
+    * init : initialize secure storage
+    * write : write item to secure storage
+    * read : read item from secure storage
+    * delete : delete item inside secure storage
+
+Here are some functions needed to use inside `impl.c`:
+* register_device : register device on startup
+* unregister_device : unregister device
+* DECLARE_DEVICE : this must be declared inside `impl.c`
+
+
+Add new device into `hal/Makefile`:
+
+* Add device object after DEVICE_OBJS
+* Add new device build target
+```
+DEVICE_OBJS = wp7702.o emulator.o device.o mydevice.o
+export DEVICE_OBJS
+
+all: $(DEVICE_OBJS)
+
+mydevice.o: device.o
+    $(MAKE) -C ../devices/mydevice
+// other device
+```
+
+Implement new device which is created under devices directory, and edit the Makefile. The example device is named as mydevice:
+```
+all: mydevice.o
+mydevice.o: impl.c
+        $(CC) $(CFLAGS) $(INCLUDES) -c $^ -o $@
+```
+`$(CC)`,`$(CFLAGS)` and `$(INCLUDES)` are provided by TA-endpoint. `CC` sets the default compiler for project, `CFLAGS` are the default flags to be passed to default compiler during compiling, `INCLUDES` flag includes headers inside sub project and third-party library. You can also modify these flag inside your device Makefile.
```

Comment:
$(CC)`,`$(CFLAGS)` and `$(INCLUDES)` are provided by TA-endpoint. `CC` sets the default compiler for the project. `CFLAGS` are the default flags that would be passed to default compiler during compiling time. `INCLUDES` flag includes headers inside sub project and third-party library. You can also modify these flags inside your device Makefile.

Reply:
`sub diretory` and `device Mkefile` is somehow misleading or unclear. It is better to use ither words to replace them